File: src/main/java/com/tjh/riskfactor/service/DataService.java
package com.tjh.riskfactor.service;

import lombok.RequiredArgsConstructor;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.tjh.riskfactor.entity.*;
import com.tjh.riskfactor.entity.form.*;
import com.tjh.riskfactor.repo.SaveGuardRepository;
import com.tjh.riskfactor.util.ThrowingRunnable;
import static com.tjh.riskfactor.util.Utils.readTreeAsType;
import static com.tjh.riskfactor.error.ResponseErrors.notFound;

import java.util.*;
import java.io.IOException;
import java.util.function.Function;
import static java.util.stream.Collectors.*;

@Service
@RequiredArgsConstructor
public class DataService {

    private final TaskService tasks;
    private final UserService users;
    private final GroupService groups;

    private final SaveGuardRepository guards;

    private static final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());

    @Transactional
    public void reloadTask() throws IOException {
        tasks.drop();
        guards.deleteById(1);
        guarded(this::loadTasks, 1);
    }

    @Transactional
    public void reloadUsers() throws IOException {
        groups.drop(); users.drop();
        guards.deleteById(0);
        guarded(this::loadUsers, 0);
    }

    @Transactional
    public void init() throws Exception {
        guarded(this::loadUsers, 0);
        guarded(this::loadTasks, 1);
    }

    private void loadTasks() throws IOException {
        var type = new TypeReference<Task>() {};
        try(var is = TypeReference.class.getResourceAsStream("/data/task.yml")) {
            var task = mapper.readValue(is, type);

            var group = groups.groupWithName(task.getCenter())
                     .orElseThrow(() -> notFound("group", task.getCenter()));
            var sections = this.tasks.saveSections(task.getSections().stream().map(this::prepareSection));
            this.tasks.save(task.setGroup(group).setSections(sections));
        }
    }

    private void loadUsers() throws IOException {
        var userListType = new TypeReference<List<User>>() {};
        var groupListType = new TypeReference<List<Group>>() {};

        try(var is = ObjectMapper.class.getResourceAsStream("/data/user.yml")) {
            var node = mapper.readTree(is); assert node.isObject();
            var root = (ObjectNode)node;

            List<User> userList = readTreeAsType(mapper, root.get("users"), userListType);
            List<Group> groupList = readTreeAsType(mapper, root.get("groups"), groupListType);

            var groups = this.groups.saveAll(groupList).stream()
                         .collect(toMap(Group::getName, g -> g));

            this.users.saveAll(userList.stream().map(
                u -> users.encoded(u.setGroup(groups.get(u.getGroupName())))
            ));
        }
    }

    private Question prepareQuestion(Question q) {
        if(q.getList() != null) {
            var questions = this.tasks.saveQuestions(
                q.getList().stream().peek(this::prepareQuestion));
            return q.setList(questions);
        }
        return q;
    }

    private Section prepareSection(Section section) {
        if(section.getQuestions() != null) {
            var questions = this.tasks.saveQuestions(
                section.getQuestions().stream().map(this::prepareQuestion));
            section.setQuestions(questions);
        }
        return section;
    }

    private <E extends Throwable> void guarded(ThrowingRunnable<E> runnable, int id) throws E {
        if(guards.existsById(id))
            return;
        runnable.run();
        guards.insert(id);
    }

}

File: backend/src/main/kotlin/com/tjh/riskfactor/service/DataService.kt
package com.tjh.riskfactor.service

import com.fasterxml.jackson.core.TreeNode
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder

import com.tjh.riskfactor.entity.*
import com.tjh.riskfactor.entity.form.*
import com.tjh.riskfactor.repo.SaveGuardRepository

@Service
class DataService(
    private val tasks: TaskService,
    private val users: UserService,
    private val groups: GroupService,
    private val questions: QuestionService,
    private val guards: SaveGuardRepository,
    builder: Jackson2ObjectMapperBuilder
) {

    private val mapper = builder.factory(YAMLFactory()).build<ObjectMapper>()

    @Transactional
    fun reloadTask() {
        tasks.drop(); guards.deleteById(0); loadTasks()
    }

    @Transactional
    fun reloadUsers() {
        groups.drop(); users.drop()
        guards.deleteById(1); loadUsers()
    }

    @Transactional
    fun init() {
        guarded(this::loadUsers, 1)
        guarded(this::loadTasks, 0)
    }

    @Transactional
    fun loadTasks() {
        val type = object: TypeReference<List<Task>>() {}

        javaClass.getResourceAsStream("/data/task.yml").use { mapper.readValue(it, type) }.forEach { task ->
            val refQuestions = mutableMapOf<String, Int>()
            val replaceQuestions = mutableListOf<Int>()
            val taskQuestions = mutableMapOf<Int, Question>()

            fun traverseQuestion(q: Question): Question {
                val empty = questions.emptyObject()
                taskQuestions[empty.id] = q
                q.list?.forEach {
                    taskQuestions[]
                }
                return empty
            }

        }
    }

    @Transactional
    fun loadUsers() {
        val userListType = object: TypeReference<List<User>>() {}
        val groupListType = object: TypeReference<List<Group>>() {}

        javaClass.getResourceAsStream("/data/user.yml").use { stream ->
            val node = mapper.readTree(stream); assert(node.isObject)
            val root = node as ObjectNode

            val userList = readTreeAs(root.get("users"), userListType)
            val groupList = readTreeAs(root.get("groups"), groupListType)

            val groups = this.groups.saveAll(groupList).associateBy { it.name }
            this.users.saveAll(userList.map {
                users.encoded(it.apply { group = groups[groupName] ?: throw Exception("group $groupName not found") })
            })
        }
    }

    private fun guarded(func: () -> Unit, guard: Int) {
        if(guards.existsById(guard))
            return
        func(); guards.insert(guard)
    }

    private fun <T> readTreeAs(tree: TreeNode, type: TypeReference<T>): T = mapper.readValue(
        mapper.treeAsTokens(tree), mapper.typeFactory.constructType(type)
    )
}
